package com.teecoin.feature.walletSystem.sendmoney;

import android.text.InputFilter;
import android.text.Spanned;

import com.teecoin.utils.TCLog;

import java.io.UnsupportedEncodingException;

public class ByteLengthFilter implements InputFilter {

    private String mCharset;
    private int mMaxByte;

    public ByteLengthFilter(int maxbyte, String charset) {
        this.mMaxByte = maxbyte;
        this.mCharset = charset;
    }

    public CharSequence filter(CharSequence source, int start, int end, Spanned dest, int dstart, int dend) {
        String expected = "";
        expected += dest.subSequence(0, dstart);
        expected += source.subSequence(start, end);
        expected += dest.subSequence(dend, dest.length());
        int keep = calculateMaxLength(expected) - (dest.length() - (dend - dstart));
        if (keep < 0) {
            keep = 0;
        }
        int Rekeep = plusMaxLength(dest.toString(), source.toString(), start);

        if (keep <= 0 && Rekeep <= 0) {
            return "";

        } else if (keep >= end - start) {
            return null;
        } else {
            if (dest.length() == 0 && Rekeep <= 0) {
                return source.subSequence(start, start + keep);
            } else if (Rekeep <= 0) {
                return source.subSequence(start, start + (source.length() - 1));
            } else {
                return source.subSequence(start, start + Rekeep);
            }
        }
    }

    private int plusMaxLength(String expected, String source, int start) {
        int keep = source.length();
        int maxByte = mMaxByte - getByteLength(expected.toString());

        TCLog.d("hung source=" + source);
        while (getByteLength(source.subSequence(start, start + keep).toString()) > maxByte) {
            keep--;
            if (keep < 0)
                return 0;
        }
        return keep;
    }

    private int calculateMaxLength(String expected) {
        int expectedByte = getByteLength(expected);
        if (expectedByte == 0) {
            return 0;
        }
        return mMaxByte - (getByteLength(expected) - expected.length());
    }

    private int getByteLength(String str) {
        try {
            return str.getBytes(mCharset).length;
        } catch (UnsupportedEncodingException e) {

        }
        return 0;
    }


}